File: Capitulo3_VRCardboard - JAVA/app/src/main/java/com/upv/pm_2022/sep_dic/capitulo3_vrcardboard/CardboardDeviceParams.java
package com.upv.pm_2022.sep_dic.capitulo3_vrcardboard;

import android.net.*;
import android.util.*;
import java.nio.*;
import com.google.protobuf.nano.*;
import java.io.*;
import android.nfc.*;

public class CardboardDeviceParams{
    private static final String TAG = "CardboardDeviceParams";
    private static final String HTTP_SCHEME = "http";
    private static final String URI_HOST_GOOGLE_SHORT = "g.co";
    private static final String URI_HOST_GOOGLE = "google.com";
    private static final String URI_PATH_CARDBOARD_HOME = "cardboard";
    private static final String URI_PATH_CARDBOARD_CONFIG = "cardboard/cfg";
    private static final String URI_SCHEME_LEGACY_CARDBOARD = "cardboard";
    private static final String URI_HOST_LEGACY_CARDBOARD = "v1.0.0";
    private static final Uri URI_ORIGINAL_CARDBOARD_NFC;
    private static final Uri URI_ORIGINAL_CARDBOARD_QR_CODE;
    private static final String URI_KEY_PARAMS = "p";
    private static final int STREAM_SENTINEL = 894990891;
    private static final String DEFAULT_VENDOR = "Google, Inc.";
    private static final String DEFAULT_MODEL = "Cardboard v1";
    private static final float DEFAULT_INTER_LENS_DISTANCE = 0.06f;
    private static final float DEFAULT_VERTICAL_DISTANCE_TO_LENS_CENTER = 0.035f;
    private static final float DEFAULT_SCREEN_TO_LENS_DISTANCE = 0.042f;
    private String mVendor;
    private String mModel;
    private float mInterLensDistance;
    private float mVerticalDistanceToLensCenter;
    private float mScreenToLensDistance;
    private FieldOfView mLeftEyeMaxFov;
    private boolean mHasMagnet;
    private Distortion mDistortion;
    
    public CardboardDeviceParams() {
        super();
        this.setDefaultValues();
    }
    
    public CardboardDeviceParams(final CardboardDeviceParams params) {
        super();
        this.copyFrom(params);
    }
    
    public CardboardDeviceParams(final CardboardDevice.DeviceParams params) {
        super();
        this.setDefaultValues();
        if (params == null) {
            return;
        }
        this.mVendor = params.getVendor();
        this.mModel = params.getModel();
        this.mInterLensDistance = params.getInterLensDistance();
        this.mVerticalDistanceToLensCenter = params.getTrayBottomToLensHeight();
        this.mScreenToLensDistance = params.getScreenToLensDistance();
        this.mLeftEyeMaxFov = FieldOfView.parseFromProtobuf(params.leftEyeFieldOfViewAngles);
        if (this.mLeftEyeMaxFov == null) {
            this.mLeftEyeMaxFov = new FieldOfView();
        }
        this.mDistortion = Distortion.parseFromProtobuf(params.distortionCoefficients);
        if (this.mDistortion == null) {
            this.mDistortion = new Distortion();
        }
        this.mHasMagnet = params.getHasMagnet();
    }
    
    public static boolean isOriginalCardboardDeviceUri(final Uri uri) {
        return CardboardDeviceParams.URI_ORIGINAL_CARDBOARD_QR_CODE.equals((Object)uri) || (CardboardDeviceParams.URI_ORIGINAL_CARDBOARD_NFC.getScheme().equals(uri.getScheme()) && CardboardDeviceParams.URI_ORIGINAL_CARDBOARD_NFC.getAuthority().equals(uri.getAuthority()));
    }
    
    private static boolean isCardboardDeviceUri(final Uri uri) {
        return HTTP_SCHEME.equals(uri.getScheme()) && URI_HOST_GOOGLE.equals(uri.getAuthority()) && URI_PATH_CARDBOARD_CONFIG.equals(uri.getPath());
    }
    
    public static boolean isCardboardUri(final Uri uri) {
        return isOriginalCardboardDeviceUri(uri) || isCardboardDeviceUri(uri);
    }
    
    public static CardboardDeviceParams createFromUri(final Uri uri) {
        if (uri == null) {
            return null;
        }
        if (isOriginalCardboardDeviceUri(uri)) {
            Log.d("CardboardDeviceParams", "URI recognized as original cardboard device.");
            final CardboardDeviceParams deviceParams = new CardboardDeviceParams();
            deviceParams.setDefaultValues();
            return deviceParams;
        }
        if (!isCardboardDeviceUri(uri)) {
            Log.w("CardboardDeviceParams", String.format("URI \"%s\" not recognized as cardboard device.", uri));
            return null;
        }
        CardboardDevice.DeviceParams params = null;
        final String paramsEncoded = uri.getQueryParameter(URI_KEY_PARAMS);
        if (paramsEncoded != null) {
            try {
                final byte[] bytes = Base64.decode(paramsEncoded, 11);
                params = (CardboardDevice.DeviceParams)MessageNano.mergeFrom((MessageNano)new CardboardDevice.DeviceParams(), bytes);
                Log.d("CardboardDeviceParams", "Read cardboard params from URI.");
            }
            catch (Exception e) {
                final String s = "CardboardDeviceParams";
                final String s2 = "Parsing cardboard parameters from URI failed: ";
                final String value = String.valueOf(e.toString());
                Log.w(s, (value.length() != 0) ? s2.concat(value) : new String(s2));
            }
        }
        else {
            Log.w("CardboardDeviceParams", "No cardboard parameters in URI.");
        }
        return new CardboardDeviceParams(params);
    }
    
    public static CardboardDeviceParams createFromInputStream(final InputStream inputStream) {
        if (inputStream == null) {
            return null;
        }
        try {
            final ByteBuffer header = ByteBuffer.allocate(8);
            if (inputStream.read(header.array(), 0, header.array().length) == -1) {
                Log.e("CardboardDeviceParams", "Error parsing param record: end of stream.");
                return null;
            }
            final int sentinel = header.getInt();
            final int length = header.getInt();
            if (sentinel != STREAM_SENTINEL) {
                Log.e("CardboardDeviceParams", "Error parsing param record: incorrect sentinel.");
                return null;
            }
            final byte[] protoBytes = new byte[length];
            if (inputStream.read(protoBytes, 0, protoBytes.length) == -1) {
                Log.e("CardboardDeviceParams", "Error parsing param record: end of stream.");
                return null;
            }
            return new CardboardDeviceParams((CardboardDevice.DeviceParams)MessageNano.mergeFrom((MessageNano)new CardboardDevice.DeviceParams(), protoBytes));
        }
        catch (InvalidProtocolBufferNanoException e) {
            final String s = "CardboardDeviceParams";
            final String s2 = "Error parsing protocol buffer: ";
            final String value = String.valueOf(e.toString());
            Log.w(s, (value.length() != 0) ? s2.concat(value) : new String(s2));
        }
        catch (IOException e2) {
            final String s3 = "CardboardDeviceParams";
            final String s4 = "Error reading Cardboard parameters: ";
            final String value2 = String.valueOf(e2.toString());
            Log.w(s3, (value2.length() != 0) ? s4.concat(value2) : new String(s4));
        }
        return null;
    }
    
    public boolean writeToOutputStream(final OutputStream outputStream) {
        try {
            final byte[] paramBytes = this.toByteArray();
            final ByteBuffer header = ByteBuffer.allocate(8);
            header.putInt(STREAM_SENTINEL);
            header.putInt(paramBytes.length);
            outputStream.write(header.array());
            outputStream.write(paramBytes);
            return true;
        }
        catch (IOException e) {
            final String s = "CardboardDeviceParams";
            final String s2 = "Error writing Cardboard parameters: ";
            final String value = String.valueOf(e.toString());
            Log.w(s, (value.length() != 0) ? s2.concat(value) : new String(s2));
            return false;
        }
    }
    
    public static CardboardDeviceParams createFromNfcContents(final NdefMessage tagContents) {
        if (tagContents == null) {
            Log.w("CardboardDeviceParams", "Could not get contents from NFC tag.");
            return null;
        }
        for (final NdefRecord record : tagContents.getRecords()) {
            final CardboardDeviceParams params = createFromUri(record.toUri());
            if (params != null) {
                return params;
            }
        }
        return null;
    }

    private byte[] toByteArray() {
        final CardboardDevice.DeviceParams params = new CardboardDevice.DeviceParams();
        params.setVendor(this.mVendor);
        params.setModel(this.mModel);
        params.setInterLensDistance(this.mInterLensDistance);
        params.setTrayBottomToLensHeight(this.mVerticalDistanceToLensCenter);
        params.setScreenToLensDistance(this.mScreenToLensDistance);
        params.leftEyeFieldOfViewAngles = this.mLeftEyeMaxFov.toProtobuf();
        params.distortionCoefficients = this.mDistortion.toProtobuf();
        if (this.mHasMagnet) {
            params.setHasMagnet(this.mHasMagnet);
        }
        return MessageNano.toByteArray((MessageNano)params);
    }
    
    public float getInterLensDistance() {
        return this.mInterLensDistance;
    }
    
    public float getVerticalDistanceToLensCenter() {
        return this.mVerticalDistanceToLensCenter;
    }
    
    public float getScreenToLensDistance() {
        return this.mScreenToLensDistance;
    }
    
    public Distortion getDistortion() {
        return this.mDistortion;
    }
    
    public FieldOfView getLeftEyeMaxFov() {
        return this.mLeftEyeMaxFov;
    }
    
    @Override
    public boolean equals(final Object other) {
        if (other == null) {
            return false;
        }
        if (other == this) {
            return true;
        }
        if (!(other instanceof CardboardDeviceParams)) {
            return false;
        }
        final CardboardDeviceParams o = (CardboardDeviceParams)other;
        return this.mVendor.equals(o.mVendor) && this.mModel.equals(o.mModel) && this.mInterLensDistance == o.mInterLensDistance && this.mVerticalDistanceToLensCenter == o.mVerticalDistanceToLensCenter && this.mScreenToLensDistance == o.mScreenToLensDistance && this.mLeftEyeMaxFov.equals(o.mLeftEyeMaxFov) && this.mDistortion.equals(o.mDistortion) && this.mHasMagnet == o.mHasMagnet;
    }
    
    private void setDefaultValues() {
        this.mVendor = DEFAULT_VENDOR;
        this.mModel = DEFAULT_MODEL;
        this.mInterLensDistance = DEFAULT_INTER_LENS_DISTANCE;
        this.mVerticalDistanceToLensCenter = DEFAULT_VERTICAL_DISTANCE_TO_LENS_CENTER;
        this.mScreenToLensDistance = DEFAULT_SCREEN_TO_LENS_DISTANCE;
        this.mLeftEyeMaxFov = new FieldOfView();
        this.mHasMagnet = true;
        this.mDistortion = new Distortion();
    }
    
    private void copyFrom(final CardboardDeviceParams params) {
        this.mVendor = params.mVendor;
        this.mModel = params.mModel;
        this.mInterLensDistance = params.mInterLensDistance;
        this.mVerticalDistanceToLensCenter = params.mVerticalDistanceToLensCenter;
        this.mScreenToLensDistance = params.mScreenToLensDistance;
        this.mLeftEyeMaxFov = new FieldOfView(params.mLeftEyeMaxFov);
        this.mHasMagnet = params.mHasMagnet;
        this.mDistortion = new Distortion(params.mDistortion);
    }
    
    static {
        URI_ORIGINAL_CARDBOARD_NFC = new Uri.Builder().scheme(URI_SCHEME_LEGACY_CARDBOARD).authority(URI_HOST_LEGACY_CARDBOARD).build();
        URI_ORIGINAL_CARDBOARD_QR_CODE = new Uri.Builder().scheme(HTTP_SCHEME).authority(URI_HOST_GOOGLE_SHORT).appendEncodedPath(URI_PATH_CARDBOARD_HOME).build();
    }
}

File: VRCardboardKotlin/app/src/main/java/com/vrcardboardkotlin/CardboardDeviceParams.kt
package com.vrcardboardkotlin

import android.net.Uri
import android.nfc.NdefMessage
import android.util.Base64
import android.util.Log
import com.google.protobuf.nano.InvalidProtocolBufferNanoException
import com.google.protobuf.nano.MessageNano
import com.vrcardboardkotlin.CardboardDevice.DeviceParams
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer

class CardboardDeviceParams {
    private var mVendor: String? = null
    private var mModel: String? = null
    var interLensDistance: Float = 0f
        private set
    var verticalDistanceToLensCenter: Float = 0f
        private set
    var screenToLensDistance: Float = 0f
        private set
    var leftEyeMaxFov: FieldOfView? = null
        private set
    private var mHasMagnet = false
    var distortion: Distortion? = null
        private set

    constructor() : super() {
        this.setDefaultValues()
    }

    constructor(params: CardboardDeviceParams?) : super() {
        this.copyFrom(params)
    }

    constructor(params: DeviceParams?) : super() {
        this.setDefaultValues()
        if (params == null) {
            return
        }
        this.mVendor = params.vendor
        this.mModel = params.model
        this.interLensDistance = params.interLensDistance
        this.verticalDistanceToLensCenter = params.trayBottomToLensHeight
        this.screenToLensDistance = params.screenToLensDistance
        this.leftEyeMaxFov =
            FieldOfView.Companion.parseFromProtobuf(params.leftEyeFieldOfViewAngles)
        if (this.leftEyeMaxFov == null) {
            this.leftEyeMaxFov = FieldOfView()
        }
        this.distortion = Distortion.Companion.parseFromProtobuf(params.distortionCoefficients)
        if (this.distortion == null) {
            this.distortion = Distortion()
        }
        this.mHasMagnet = params.hasMagnet
    }

    fun writeToOutputStream(outputStream: OutputStream): Boolean {
        try {
            val paramBytes = this.toByteArray()
            val header = ByteBuffer.allocate(8)
            header.putInt(STREAM_SENTINEL)
            header.putInt(paramBytes.size)
            outputStream.write(header.array())
            outputStream.write(paramBytes)
            return true
        } catch (e: IOException) {
            val s = "CardboardDeviceParams"
            val s2 = "Error writing Cardboard parameters: "
            val value = e.toString().toString()
            Log.w(s, if ((value.length != 0)) s2 + value else java.lang.String(s2) as String)
            return false
        }
    }

    private fun toByteArray(): ByteArray {
        val params = DeviceParams()
        params.setVendor(this.mVendor)
        params.setModel(this.mModel)
        params.setInterLensDistance(this.interLensDistance)
        params.setTrayBottomToLensHeight(this.verticalDistanceToLensCenter)
        params.setScreenToLensDistance(this.screenToLensDistance)
        params.leftEyeFieldOfViewAngles = leftEyeMaxFov!!.toProtobuf()
        params.distortionCoefficients = distortion!!.toProtobuf()
        if (this.mHasMagnet) {
            params.setHasMagnet(this.mHasMagnet)
        }
        return MessageNano.toByteArray(params as MessageNano)
    }

    override fun equals(other: Any?): Boolean {
        if (other == null) {
            return false
        }
        if (other === this) {
            return true
        }
        if (other !is CardboardDeviceParams) {
            return false
        }
        val o = other
        return this.mVendor == o.mVendor && (this.mModel == o.mModel) && (this.interLensDistance == o.interLensDistance) && (this.verticalDistanceToLensCenter == o.verticalDistanceToLensCenter) && (this.screenToLensDistance == o.screenToLensDistance) && (this.leftEyeMaxFov == o.leftEyeMaxFov) && (this.distortion == o.distortion) && (this.mHasMagnet == o.mHasMagnet)
    }

    private fun setDefaultValues() {
        this.mVendor = DEFAULT_VENDOR
        this.mModel = DEFAULT_MODEL
        this.interLensDistance = DEFAULT_INTER_LENS_DISTANCE
        this.verticalDistanceToLensCenter = DEFAULT_VERTICAL_DISTANCE_TO_LENS_CENTER
        this.screenToLensDistance = DEFAULT_SCREEN_TO_LENS_DISTANCE
        this.leftEyeMaxFov = FieldOfView()
        this.mHasMagnet = true
        this.distortion = Distortion()
    }

    private fun copyFrom(params: CardboardDeviceParams?) {
        this.mVendor = params!!.mVendor
        this.mModel = params.mModel
        this.interLensDistance = params.interLensDistance
        this.verticalDistanceToLensCenter = params.verticalDistanceToLensCenter
        this.screenToLensDistance = params.screenToLensDistance
        this.leftEyeMaxFov = FieldOfView(params.leftEyeMaxFov)
        this.mHasMagnet = params.mHasMagnet
        this.distortion = Distortion(params.distortion)
    }

    companion object {
        private const val TAG = "CardboardDeviceParams"
        private const val HTTP_SCHEME = "http"
        private const val URI_HOST_GOOGLE_SHORT = "g.co"
        private const val URI_HOST_GOOGLE = "google.com"
        private const val URI_PATH_CARDBOARD_HOME = "cardboard"
        private const val URI_PATH_CARDBOARD_CONFIG = "cardboard/cfg"
        private const val URI_SCHEME_LEGACY_CARDBOARD = "cardboard"
        private const val URI_HOST_LEGACY_CARDBOARD = "v1.0.0"
        private val URI_ORIGINAL_CARDBOARD_NFC: Uri = Uri.Builder().scheme(
            URI_SCHEME_LEGACY_CARDBOARD
        ).authority(URI_HOST_LEGACY_CARDBOARD).build()
        private val URI_ORIGINAL_CARDBOARD_QR_CODE: Uri =
            Uri.Builder().scheme(HTTP_SCHEME).authority(URI_HOST_GOOGLE_SHORT).appendEncodedPath(
                URI_PATH_CARDBOARD_HOME
            ).build()
        private const val URI_KEY_PARAMS = "p"
        private const val STREAM_SENTINEL = 894990891
        private const val DEFAULT_VENDOR = "Google, Inc."
        private const val DEFAULT_MODEL = "Cardboard v1"
        private const val DEFAULT_INTER_LENS_DISTANCE = 0.06f
        private const val DEFAULT_VERTICAL_DISTANCE_TO_LENS_CENTER = 0.035f
        private const val DEFAULT_SCREEN_TO_LENS_DISTANCE = 0.042f
        fun isOriginalCardboardDeviceUri(uri: Uri): Boolean {
            return URI_ORIGINAL_CARDBOARD_QR_CODE == uri as Any || (URI_ORIGINAL_CARDBOARD_NFC.scheme == uri.scheme && URI_ORIGINAL_CARDBOARD_NFC.authority == uri.authority)
        }

        private fun isCardboardDeviceUri(uri: Uri): Boolean {
            return HTTP_SCHEME == uri.scheme && URI_HOST_GOOGLE == uri.authority && URI_PATH_CARDBOARD_CONFIG == uri.path
        }

        fun isCardboardUri(uri: Uri): Boolean {
            return isOriginalCardboardDeviceUri(uri) || isCardboardDeviceUri(uri)
        }

        fun createFromUri(uri: Uri?): CardboardDeviceParams? {
            if (uri == null) {
                return null
            }
            if (isOriginalCardboardDeviceUri(uri)) {
                Log.d("CardboardDeviceParams", "URI recognized as original cardboard device.")
                val deviceParams = CardboardDeviceParams()
                deviceParams.setDefaultValues()
                return deviceParams
            }
            if (!isCardboardDeviceUri(uri)) {
                Log.w(
                    "CardboardDeviceParams",
                    String.format("URI \"%s\" not recognized as cardboard device.", uri)
                )
                return null
            }
            var params: DeviceParams? = null
            val paramsEncoded = uri.getQueryParameter(URI_KEY_PARAMS)
            if (paramsEncoded != null) {
                try {
                    val bytes = Base64.decode(paramsEncoded, 11)
                    params =
                        MessageNano.mergeFrom(DeviceParams() as MessageNano, bytes) as DeviceParams
                    Log.d("CardboardDeviceParams", "Read cardboard params from URI.")
                } catch (e: Exception) {
                    val s = "CardboardDeviceParams"
                    val s2 = "Parsing cardboard parameters from URI failed: "
                    val value = e.toString().toString()
                    Log.w(
                        s,
                        if ((value.length != 0)) s2 + value else java.lang.String(s2) as String
                    )
                }
            } else {
                Log.w("CardboardDeviceParams", "No cardboard parameters in URI.")
            }
            return CardboardDeviceParams(params)
        }

        fun createFromInputStream(inputStream: InputStream?): CardboardDeviceParams? {
            if (inputStream == null) {
                return null
            }
            try {
                val header = ByteBuffer.allocate(8)
                if (inputStream.read(header.array(), 0, header.array().size) == -1) {
                    Log.e("CardboardDeviceParams", "Error parsing param record: end of stream.")
                    return null
                }
                val sentinel = header.getInt()
                val length = header.getInt()
                if (sentinel != STREAM_SENTINEL) {
                    Log.e(
                        "CardboardDeviceParams",
                        "Error parsing param record: incorrect sentinel."
                    )
                    return null
                }
                val protoBytes = ByteArray(length)
                if (inputStream.read(protoBytes, 0, protoBytes.size) == -1) {
                    Log.e("CardboardDeviceParams", "Error parsing param record: end of stream.")
                    return null
                }
                return CardboardDeviceParams(
                    MessageNano.mergeFrom(
                        DeviceParams() as MessageNano,
                        protoBytes
                    ) as DeviceParams
                )
            } catch (e: InvalidProtocolBufferNanoException) {
                val s = "CardboardDeviceParams"
                val s2 = "Error parsing protocol buffer: "
                val value = e.toString().toString()
                Log.w(s, if ((value.length != 0)) s2 + value else java.lang.String(s2) as String)
            } catch (e2: IOException) {
                val s3 = "CardboardDeviceParams"
                val s4 = "Error reading Cardboard parameters: "
                val value2 = e2.toString().toString()
                Log.w(s3, if ((value2.length != 0)) s4 + value2 else java.lang.String(s4) as String)
            }
            return null
        }

        fun createFromNfcContents(tagContents: NdefMessage?): CardboardDeviceParams? {
            if (tagContents == null) {
                Log.w("CardboardDeviceParams", "Could not get contents from NFC tag.")
                return null
            }
            for (record in tagContents.records) {
                val params = createFromUri(record.toUri())
                if (params != null) {
                    return params
                }
            }
            return null
        }
    }
}
